libressl-build/share/man/man3/RAND.3
.\" Automatically generated by Pod::Man 2.28 (Pod::Simple 3.28)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
.    ds C`
.    ds C'
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.\"
.\" Avoid warning from groff about undefined register 'F'.
.de IX
..
.nr rF 0
.if \n(.g .if rF .nr rF 1
.if (\n(rF:(\n(.g==0)) \{
.    if \nF \{
.        de IX
.        tm Index:\\$1\t\\n%\t"\\$2"
..
.        if !\nF==2 \{
.            nr % 0
.            nr F 2
.        \}
.    \}
.\}
.rr rF
.\" ========================================================================
.\"
.IX Title "RAND 3"
.TH RAND 3 "2017-01-09" "LibreSSL " "LibreSSL"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "NAME"
RAND \- pseudo\-random number generator
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
.Vb 1
\& #include <openssl/rand.h>
\&
\& int  RAND_bytes(unsigned char *buf, int num);
\& int  RAND_pseudo_bytes(unsigned char *buf, int num);
.Ve
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
These functions give access to the systems cryptographically secure
pseudo-random number generator (\s-1PRNG\s0). It is used by other library functions
for example to generate random keys, and applications can use it when they
need randomness.
.PP
\&\fIRAND_bytes\fR\|(3) describes how to obtain random data from the
\&\s-1PRNG.\s0
.SH "INTERNALS"
.IX Header "INTERNALS"
The \fIRAND_SSLeay()\fR method implements a \s-1PRNG\s0 based on the systems'
\&\fIarc4random_buf\fR\|(3) random number generator.
.SH "SEE ALSO"
.IX Header "SEE ALSO"
\&\fIBN_rand\fR\|(3),
\&\fIRAND_bytes\fR\|(3)


Changes
Revision history for H2O.

2.3.0-beta1 TBD
	- [access-log] log connections that closed prior to sending a response #1235 (Kazuho Oku)
	- [ssl] support redis as a data store for session cache / ticket #1087 (Ichito Nagata)
	- [misc] include process id in crash backtrace #1254 (Kazuho Oku)

2.2.2 2017-04-23 03:24:00+0000
	- [ssl] fix OCSP stapling error when LibreSSL is used #1275 (Ian Moone)


2.2.1 2017-04-22 22:43:00+0000
	- [mruby] correct the line number reported on an exception #1239 #1251 (Ichito Nagata)
	- [mruby] retain the order of request headers sharing a single name #1271 (Kazuho Oku)
	- [ssl] fix assertion failure in `decode_ssl_input` #1264 (Kazuho Oku)
	- [ssl] fix OCSP stapling error when OpenSSL 1.1.0 is used #1270 (Kazuho Oku)
	- [libh2o] fix crash when abruptly closing an HTTP/2 connection on libuv #1250 (Kazuho Oku)
	- [libh2o] fix memory leak of `_timestamp_cache` #1255 (Kazuho Oku)
	- [doc] restore doc of `%{...}e` #1252 (Kazuho Oku)
	- [doc] fix typo suggesting using `brotli` instead of `br` #1263 (Bogdan Khomutsky)
	- [misc] fix undefined behaviors detected by ubsan #1246 (Frederik Deweerdt)

2.2.0 2017-04-05 02:18:00+0000
	- [core] add `crash-handler.wait-pipe-close` parameter #1092 (Frederik Deweerdt)
	- [core] introduce an option to bypass the `server` header sent from upstream #1226 (Frederik Deweerdt)
	- [core] apply global- and host-level configuration to requests not applicable to any of the path-level configurations #1231 (Kazuho Oku)
	- [access-log] add `%{remote}p` for logging the remote port #1166 (Kazuho Oku)
	- [access-log] add support for JSON-style escapes and null #1208 (Kazuho Oku)
	- [access-log] add specifier for logging per-request environment variables #1221 (Yannick Koechlin)
	- [access-log] add support for `<`, `>` modifiers for logging either the original or the final response #1238 (Kazuho Oku)
	- [access-log] do not emit request-total-time twice #1017 (Kazuho Oku)
	- [fastcgi] fix a bug that closes the FastCGI listener socket during startup #1203 (Frederik Deweerdt)
	- [file] add directive for serving gzipped files, decompressing them on-the-fly #1140 (Ichito Nagata)
	- [headers] fix buffer overrun during startup #1180 (Frederik Deweerdt)
	- [http1][proxy] preserve the cases of characters used in header names #1194 (Frederik Deweerdt)
	- [http1][proxy] fix undefined behavior in HTTP/1 parser #1189 (Frederik Deweerdt)
	- [http1] stop reading from socket after sending 400 to avoid the risk of assertion failure #1223 (Frederik Deweerdt)
	- [http2] recognize `x-http2-push-only` attribute on `link` header #1169 (Frederik Deweerdt)
	- [http2] add optional timeout for closing connections upon graceful shutdown #1108 (Frederik Deweerdt)
	- [http2] do not ack an acked PING frame #1175 (Moto Ishisawa)
	- [http2] reject requests exceeding the maximum allowed size more efficiently #1183 (Frederik Deweerdt)
	- [mruby] remove dependenty to mkmf #1197 (Yuki Kurihara)
	- [mruby] correct the line number reported on an exception #1239 (Ichito Nagata)
	- [proxy] add directives for tweaking headers sent to upstream #1126 (Justin Zhu)
	- [proxy] retain case-sensitivity of unix socket paths #1131 (Frederik Deweerdt)
	- [proxy] add directive for controlling the `via` request header #1225 (Frederik Deweerdt)
	- [ssl] add directive for logging session ID #1164 (Yannick Koechlin)
	- [ssl] add support for TLS 1.3 draft-18 #1204 (Kazuho Oku)
	- [ssl] stop evicting session entries in memcached when they are removed from internal cache #1185 (Ichito Nagata)
	- [ssl] fix crash when a secp384r1, secp521r1 certificate is used with TLS 1.3 #1214 (Kazuho Oku)
	- [ssl] fix build failure with OpenSSL 1.1.0 #1216 (Kazuho Oku)
	- [ssl] add doc for `handshake-timeout` #1233 (Kazuho Oku)
	- [status] fix race condition during start-up #1242 (Frederik Deweerdt)
	- [libh2o] implement `h2o_evloop_destroy` #1200 (kazan417)
	- [misc] add test code for fuzzing #1174 #1182 #1191 #1192 (Frederik Deweerdt, Jonathan Foote)
	- [misc] fix issues reported by Coverity #1168 #1172 #1179 (Harrison Bowden, Frederik Deweerdt)

2.1.0 2017-01-17 23:43:27+0000
	- [core] TCP latency optimization #873 #1076 (Kazuho Oku)
	- [core] provide tag to include other YAML files from the configuration file #1022 (Ichito Nagata)
	- [core] accept sequence of mappings for path-level configuration #1042 (Ichito Nagata)
	- [core] fix broken support for TCP Fast Open in OS X #1065 (Ichito Nagata)
	- [access-log] provide directive to emit request-level errors #1075 (Kazuho Oku)
	- [access-log] emit values of all `set-cookie` headers concatenated #1161 (Kazuho Oku)
	- [fastcgi] fix connection failure when `fastcgi.spawn` is used with an uid #1119 (Kazuho Oku)
	- [file] more pre-defined MIME types #1103 (Joe Duarte)
	- [http2][proxy] recognize link rel=preload headers in interim response as a trigger to push resources #916 (Kazuho Oku)
	- [http1][http2] validate characters used in the headers #974 #1044 (Frederik Deweerdt, Kazuho Oku)
	- [http1][http2] notify error downstream when an error occurred while generating a response #1031 (Frederik Deweerdt)
	- [http1][http2] fix resource leak upon upgrade failure to HTTP/2 #1161 (Frederik Deweerdt)
	- [http2] add `http2-push-preload` directive to turn off H2 push being initiated by link rel=preload header #929 (Kazuho Oku)
	- [http2] add support for `cache-digest` header #967 #988 (Kazuho Oku)
	- [http2] drop `host` header in HTTP/2 layer #973 #998 (Frederik Deweerdt, Kazuho Oku)
	- [http2] don't use etag for calculating casper cookie #986 (Kazuho Oku)
	- [http2] add support for H2 debug state #1019 (Ichito Nagata)
	- [mruby] add dos_detector mruby handler #1013 (Ichito Nagata)
	- [mruby] add DSL for access control lists (acl) #1016 (Ichito Nagata)
	- [mruby] share mruby state and constants between handlers #1032 (Ichito Nagata)
	- [mruby] add library for address-block-based access control #1038 (Ichito Nagata)

libressl-2.4.5/apps/openssl/rand.c
/* $OpenBSD: rand.c,v 1.9 2015/10/10 22:28:51 doug Exp $ */
/* ====================================================================
 * Copyright (c) 1998-2001 The OpenSSL Project.  All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 *
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 *
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in
 *    the documentation and/or other materials provided with the
 *    distribution.
 *
 * 3. All advertising materials mentioning features or use of this
 *    software must display the following acknowledgment:
 *    "This product includes software developed by the OpenSSL Project
 *    for use in the OpenSSL Toolkit. (http://www.openssl.org/)"
 *
 * 4. The names "OpenSSL Toolkit" and "OpenSSL Project" must not be used to
 *    endorse or promote products derived from this software without
 *    prior written permission. For written permission, please contact
 *    openssl-core@openssl.org.
 *
 * 5. Products derived from this software may not be called "OpenSSL"
 *    nor may "OpenSSL" appear in their names without prior written
 *    permission of the OpenSSL Project.
 *
 * 6. Redistributions of any form whatsoever must retain the following
 *    acknowledgment:
 *    "This product includes software developed by the OpenSSL Project
 *    for use in the OpenSSL Toolkit (http://www.openssl.org/)"
 *
 * THIS SOFTWARE IS PROVIDED BY THE OpenSSL PROJECT ``AS IS'' AND ANY
 * EXPRESSED OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
 * PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE OpenSSL PROJECT OR
 * ITS CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
 * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
 * NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
 * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
 * STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED
 * OF THE POSSIBILITY OF SUCH DAMAGE.
 * ====================================================================
 *
 * This product includes cryptographic software written by Eric Young
 * (eay@cryptsoft.com).  This product includes software written by Tim
 * Hudson (tjh@cryptsoft.com).
 *
 */

#include <ctype.h>
#include <stdio.h>
#include <string.h>

#include "apps.h"

#include <openssl/bio.h>
#include <openssl/err.h>

struct {
	int base64;
	int hex;
	char *outfile;
} rand_config;

struct option rand_options[] = {
	{
		.name = "base64",
		.desc = "Perform base64 encoding on output",
		.type = OPTION_FLAG,
		.opt.flag = &rand_config.base64,
	},
	{
		.name = "hex",
		.desc = "Hexadecimal output",
		.type = OPTION_FLAG,
		.opt.flag = &rand_config.hex,
	},
	{
		.name = "out",
		.argname = "file",
		.desc = "Write to the given file instead of standard output",
		.type = OPTION_ARG,
		.opt.arg = &rand_config.outfile,
	},
	{NULL},
};

static void
rand_usage()
{
	fprintf(stderr,
	    "usage: rand [-base64 | -hex] [-out file] num\n");
	options_usage(rand_options);
}

int
rand_main(int argc, char **argv)
{
	char *num_bytes = NULL;
	int ret = 1;
	int badopt = 0;
	int num = -1;
	int i, r;
	BIO *out = NULL;

	if (single_execution) {
		if (pledge("stdio rpath wpath cpath", NULL) == -1) {
			perror("pledge");
			exit(1);
		}
	}

	memset(&rand_config, 0, sizeof(rand_config));

	if (options_parse(argc, argv, rand_options, &num_bytes, NULL) != 0) {
		rand_usage();
		return (1);
	}

	if (num_bytes != NULL) {
		r = sscanf(num_bytes, "%d", &num);
		if (r == 0 || num < 0)
			badopt = 1;
	} else
		badopt = 1;

	if (rand_config.hex && rand_config.base64)
		badopt = 1;

	if (badopt) {
		rand_usage();
		goto err;
	}

	out = BIO_new(BIO_s_file());
	if (out == NULL)
		goto err;
	if (rand_config.outfile != NULL)
		r = BIO_write_filename(out, rand_config.outfile);
	else
		r = BIO_set_fp(out, stdout, BIO_NOCLOSE | BIO_FP_TEXT);
	if (r <= 0)
		goto err;
	if (rand_config.base64) {
		BIO *b64 = BIO_new(BIO_f_base64());
		if (b64 == NULL)
			goto err;
		out = BIO_push(b64, out);
	}

	while (num > 0) {
		unsigned char buf[4096];
		int chunk;

		chunk = num;
		if (chunk > (int) sizeof(buf))
			chunk = sizeof(buf);
		arc4random_buf(buf, chunk);
		if (rand_config.hex) {
			for (i = 0; i < chunk; i++)
				BIO_printf(out, "%02x", buf[i]);
		} else
			BIO_write(out, buf, chunk);
		num -= chunk;
	}

	if (rand_config.hex)
		BIO_puts(out, "\n");
	(void) BIO_flush(out);

	ret = 0;

err:
	ERR_print_errors(bio_err);
	if (out)
		BIO_free_all(out);

	return (ret);
}
